File: python/railroaded/models/agency.py
from typing import Optional

import seared as s


@s.seared
class Agency(s.Seared):
    '''
    A GTFS dataclass model for records found in `agency.txt`. Identifies a 
    transit agency.

    Attributes:
        id (str):
            the unique ID of the transit agency
        email (Optional[str]):
            the customer service email of the transit agency
        fare_url (Optional[str]):
            the URL of a fare or ticket website for the transit agency
        lang (Optional[str]):
            the primary language used by the transit agency
        name (str):
            the name of the transit agency
        phone (Optional[str]):
            the voice telephone number for the transit agency
        timezone (str):
            the timezone where the transit agency is located
        url (str):
            the URL of the transit agency
    '''
    
    ### ATTRIBUTES ###
    # Model ID
    id: str = s.Str(data_key='agency_id', missing='')
    '''the unique ID of the transit agency'''
    
    # Required fields
    name: str = s.Str(data_key='agency_name', required=True)
    '''the name of the transit agency'''
    timezone: str = s.Str(data_key='agency_timezone', required=True)
    '''the timezone where the transit agency is located'''
    url: str = s.Str(data_key='agency_url', required=True)
    '''the URL of the transit agency'''

    # Optional fields
    email: Optional[str] = s.Str(data_key='agency_email')
    '''the customer service email of the transit agency'''
    fare_url: Optional[str] = s.Str(data_key='agency_fare_url')
    '''the URL of a fare or ticket website for the transit agency'''
    lang: Optional[str] = s.Str(data_key='agency_lang')
    '''the primary language used by the transit agency'''
    phone: Optional[str] = s.Str(data_key='agency_phone')
    '''the voice telephone number for the transit agency'''

File: typescript/src/models/agency.ts
export type GTFSAgency = {
    agency_id?: string
    agency_email?: string
    agency_fare_url?: string
    agency_lang?: string
    agency_name: string
    agency_phone?: string
    agency_timezone: string
    agency_url: string
}


export type MGTFSAgency = {
    id: string
    email?: string
    fare_url?: string
    lang?: string
    name: string
    phone?: string
    timezone: string
    url: string
}


/**
 * A GTFS model for records found in `agency.txt`. Identifies a transit agency.
 * 
 * @property id - the unique ID of the transit agency
 * @property email - the customer service email of the transit agency
 * @property fareURL - the URL of a fare or ticket website for the transit agency
 * @property lang - the primary language used by the transit agency
 * @property name - the name of the transit agency
 * @property phone - the voice telephone number for the transit agency
 * @property timezone - the timezone where the transit agency is located
 * @property url - the URL of the transit agency
 */
export class AFgency {

    id: string
    email?: string
    fareURL?: string
    lang?: string
    name: string
    phone?: string
    timezone: string
    url: string

    constructor (data: MGTFSAgency) {
        this.id = data.id
        this.email = data.email
        this.fareURL = data.fare_url
        this.lang = data.lang
        this.name = data.name
        this.phone = data.phone
        this.timezone = data.timezone
        this.url = data.url
    }

    static fromGTFS (data: GTFSAgency): MGTFSAgency {
        return {
            id: data.agency_id ?? '',
            email: data.agency_email,
            fare_url: data.agency_fare_url,
            lang: data.agency_lang,
            name: data.agency_name,
            phone: data.agency_phone,
            timezone: data.agency_timezone,
            url: data.agency_url
        }
    }

    toMGTFS (): MGTFSAgency {
        return {
            id: this.id,
            email: this.email,
            fare_url: this.fareURL,
            lang: this.lang,
            name: this.name,
            phone: this.phone,
            timezone: this.timezone,
            url: this.url
        }
    }

}